---
title: "Phage Proteome Analysis"
output: html_document
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=FALSE, warning=FALSE, message=FALSE)

library(dplyr)
library(ggplot2)
library(cowplot)
library(readr)
library(stringr)
```


We have proteomics data from three different strains of T7 after 1, 5, and 9 minutes of infection in E. coli. Each time point has 4 replicates.

**evol**: evolved phage with codon-modified gene 10

**atten**: initial codon-modified gene 10 phage

**wt**: wild-type phage

**Normalization:** Data are normalized by total protein area (T7 and E. coli) for each sample _after_ computing areas.


```{r}

# Load in replicates
# rep1 <- read_csv("data/post_processed/rep1.csv") %>% mutate(time = paste0(time, "min"))
rep2 <- read_csv("data/post_processed/rep2.csv")
rep3 <- read_csv("data/post_processed/rep3.csv")
rep4 <- read_csv("data/post_processed/rep4.csv")
rep5 <- read_csv("data/post_processed/rep5.csv")

all_data <- bind_rows(rep2, rep3, rep4, rep5) %>%
  mutate(strain = str_replace_all(strain, c("11-44" = "atten", "11-42" = "evol", "11-46" = "wt"))) %>%
  mutate(Proteins = trimws(Proteins))

# Correct wonky labeling
all_data <- mutate(all_data, time = ifelse(b_rep == 1 & time == "1min" & strain == "wt", "temp", time)) %>% 
  mutate(time = ifelse(b_rep == 1 & time == "9min" & strain == "wt", "1min", time)) %>% 
  mutate(time = ifelse(b_rep == 1 & time == "temp" & strain == "wt", "9min", time))

# Join in standard gene labels
labels <- read_csv("data/id_map.csv") %>% mutate(Accession = trimws(Accession))
all_data <- left_join(all_data, labels, by = c("Proteins" = "Accession"))
```

### Major capsid protein levels are depressed in attenuated and evolved strains

```{r}
ggplot(filter(all_data, tabasco_id == 'gp10A'), aes(x = strain, y = area_norm, group = strain, color=factor(strain), order = desc(strain))) + 
  geom_point(size = 3) +
  stat_summary(geom="errorbar", fun.y="mean", aes(ymin=..y.., ymax=..y..), size=1) +
  facet_wrap(~ time) +
  panel_border() +
  ylab("relative protein abundance") +
  theme(legend.position = "none") +
  geom_text(aes(x = strain, y = area_norm, label = b_rep), hjust = -0.5)

```

## Some phage proteins immediately downstream of gp10A are also suppressed

### Peak areas
```{r, fig.height=20}

ggplot(filter(all_data, tabasco_id %in% c('gp10A','gp11','gp12','gp13','gp14','gp15')), aes(x = strain, y = area_norm, group = strain, color=factor(strain), order = desc(strain))) + 
  geom_point(size = 3) +
  stat_summary(geom="errorbar", fun.y="mean", aes(ymin=..y.., ymax=..y..), size=1) +
  facet_grid(tabasco_id ~ time, scales = "free_y") +
  panel_border() +
  ylab("relative protein abundance") +
  theme(legend.position = "none") +
  geom_text(aes(x = strain, y = area_norm, label = b_rep), hjust = -0.5)

```

## In aggregate, phage protein levels (not including major capsid) look about the same across strains

```{r}

make_abundance_plot <- function(protein_data) {

  sums <- group_by(protein_data, strain, time, b_rep) %>% 
    summarize(area_sum2 = sum(area_norm)) %>% 
    filter(!is.na(area_sum2)) %>%
    group_by(strain, time) %>%
    summarize(area_mean = mean(area_sum2), area_sd = sd(area_sum2)/sqrt(n())) %>%
    mutate(upper_error = area_mean + area_sd, lower_error = area_mean - area_sd)

  myplot <- ggplot(sums, aes(x = strain, y = area_mean, group = strain, fill=factor(strain))) + 
    geom_bar(stat="identity") +
    geom_errorbar(aes(ymin = lower_error, ymax = upper_error), width = 0.5) +
    facet_wrap(~ time) +
    panel_border() +
    ylab('total relative protein abundance') +
    theme(legend.position = "none")
  
  return(myplot)

}

# Plot all data
plot_all_data <- make_abundance_plot(all_data)

# Plot without gp10A
plot_no10 <- make_abundance_plot(filter(all_data, 
                                        tabasco_id != 'gp10A', 
                                        tabasco_id != 'gp10B'))

plot_no10

```

### Class 3 protein levels (not including capsid) look roughly the same

```{r}

plot_class3_no10 <- make_abundance_plot(filter(all_data, 
                                               tabasco_id != 'gp10A', 
                                               tabasco_id != 'gp10B', 
                                               class == 3))

plot_class3_no10

```

Averaged across all genes and all replicates for a given time point and given strain, class 3 protein levels look similar to gp10A (just barely). I have not actually tested to see if any strain is significantly different from another across all genes (see mixed effects ANOVA below).

### Classes 1 and 2 also look roughly the same across all three strains

```{r}

plot_class12 <- make_abundance_plot(filter(all_data, class != 3))

plot_class12

```

## Mixed models: Can we control for batch effects?

Looking at the class 3 genes, there seem to be systematic batch effects between biological replicates. For example, replicate 4 might have lower abundances for a given protein in all three strains. I've attempted to model this with a mixed effects model:

```{r, echo = T}
library(lmerTest)

# Let's just look at the 9min mark without the evolved strain or gp10A
all_data2 <- filter(all_data, tabasco_id != "gp10A", time == "9min", strain != "evol", class == 3)

# Build a model with protein abundance as the response, strain and protein as fixed, and biological replicate as random
mod <- lmer(area_norm ~ strain + tabasco_id + (1 | b_rep), data = all_data2)

anova(mod)

```

According to the ANOVA test above, the protein abundances for class 3 genes (excluding gp10A) across strains are significantly different. I still have to make sure that some assumptions made by ANOVA are not being violated here.

