---
title: "Transcriptional Effects of T7 Attenuation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(readr)
library(dplyr)
library(stringr)
library(cowplot)

rna <- read_csv("./data/all_rna.csv") %>%
  # Replace strain names with something more useful
  mutate(strain = str_replace_all(strain, c("11-42" = "evolved", "11-44" = "recoded", "11-46" = "wildtype")))

```

## What proportion of transcripts are from T7?

At first glance, the abundance of T7 transcripts across strains and time points looks wildly different:

```{r fig.width = 9}

phagerna <- rna %>% 
  filter(org == "phage") %>%
  group_by(b_rep, time, strain) %>%
  summarize(tpm = sum(tpm))
  
rna_raw_plot <- ggplot(phagerna, aes(x = strain, y = tpm, fill = factor(b_rep))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid( ~ time)

rna_raw_plot

save_plot("figures/rna_raw.pdf", rna_raw_plot, base_aspect_ratio = 2)

```

These differences in TPM (transcripts per million), are driven by differences in rRNA levels:

```{r fig.width = 9}

phagerna <- rna %>%
  group_by(b_rep, strain, time, gene_type) %>%
  summarize(counts = sum(tpm))

ribo_rna <- ggplot(phagerna, aes(x = strain, y = counts, fill = factor(gene_type))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid(b_rep ~ time)

save_plot("figures/ribo_rna.pdf", ribo_rna, base_aspect_ratio = 2)

```

To account for different rRNA levels, I took the raw read counts and re-normalized them to the total mRNA content. This makes the phage transcript levels at least look more sane.

```{r fig.width = 9}
phagerna <- rna %>% 
  filter(gene_type == "CDS") %>%
  group_by(b_rep, strain, time) %>%
  mutate(total_count = sum(est_counts), est_counts_renorm = est_counts/total_count) %>% 
  group_by(b_rep, strain, time, org) %>%
  summarize(counts = sum(est_counts_renorm))
  
ggplot(phagerna, aes(x = strain, y = counts, fill = factor(b_rep))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid(org ~ time)
```

Lastly, lets look at gene 10A:

```{r}

phagerna <- filter(rna, gene_type == "CDS") %>%
  group_by(b_rep, strain, time) %>%
  mutate(total_count = sum(est_counts), est_counts_renorm = est_counts/total_count) %>%
  filter(gene_name == "T7p45")

ggplot(phagerna, aes(x = strain, y = est_counts_renorm, fill = factor(b_rep))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid(org ~ time)

```

...And T7 polymerase:

```{r}

phagerna <- filter(rna, gene_type == "CDS") %>%
  group_by(b_rep, strain, time) %>%
  mutate(total_count = sum(est_counts), est_counts_renorm = est_counts/total_count) %>%
  filter(gene_name == "T7p07")

ggplot(phagerna, aes(x = strain, y = est_counts_renorm, fill = factor(b_rep))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid(org ~ time)

```

Any individual genes still look somewhat noisy. This could have to do with the timing, or maybe a more sophisticated normalization method is required. Next, I plan to re-run kallisto with a reference transcriptome that doesn't contain any rRNA genes to see if kallisto's normalization is better than simply summing the estimated counts as I did above.

## Mapping to a transcriptome without ribosomal RNA

I re-ran kallisto with a reference transcriptome that doesn't contain any rRNA genes. I also ran it with the experimentally derived mean fragment length of 450 and sd of 250. This causes short transcripts to drop out, but these are transcripts that can't be reliably quantitated given the fragment length.

Let's look at the overall proportion of transcripts between T7 and E. coli.

```{r}

strain_colors <- c(
  "recoded" = "orange",
  "evolved" = "lightblue",
  "wildtype" = "lightgreen"
)

rna2 <- read_csv("./data/all_rna_no_ribo_trna_merge.csv") %>%
  # Replace strain names with something more useful
  mutate(strain = str_replace_all(strain, c("11-42" = "evolved", "11-44" = "recoded", "11-46" = "wildtype"))) %>%
  filter(gene_type != "tRNA") %>%
  mutate(tpm = tpm/1e6)

rna2$strain <- factor(rna2$strain, levels = c("recoded", "evolved", "wildtype"))
rna2$time <- factor(rna2$time, labels = c("1min" = "1 minute", "5min" = "5 minutes", "9min" = "9 minutes"))

phagerna2 <- rna2 %>% 
  group_by(b_rep, time, strain, org) %>%
  summarize(tpm = sum(tpm))

phagerna2$org <- factor(phagerna2$org, labels = c("ecoli" = "E. coli", "phage" = "T7"))
  
rna_no_ribo <- ggplot(phagerna2, aes(x = strain, y = tpm, group=strain)) + 
  stat_summary(geom="bar", fun.y="mean", aes(fill = strain)) +
  scale_fill_manual(values = strain_colors) +
  geom_point(aes(group=b_rep)) +
  geom_line(aes(group=b_rep)) +
  facet_grid(org ~ time) +
  panel_border() +
  ylab("relative mRNA abundance") +
  theme(legend.position = "none")

rna_no_ribo

save_plot("figures/rna_no_ribo.pdf", rna_no_ribo, base_height = 5, base_width = 8)
```

Gene 10A transcripts and gene 1 transcripts:

```{r}
phagerna2 <- filter(rna2, gene_name == "T7p45")

gp10_plot <- ggplot(phagerna2, aes(x = strain, y = tpm, group = b_rep)) + 
  stat_summary(geom="bar", fun.y = "mean", aes(group = strain, fill = strain)) +
  geom_point() +
  geom_line(aes(group = b_rep)) + 
  scale_fill_manual(values = strain_colors) +
  facet_grid(~ time) +
  ylab("relative mRNA abundance") +
  panel_border() +
  theme(legend.position = "none")

gp10_plot

save_plot("figures/gp10_rna.pdf", gp10_plot, base_aspect_ratio = 2)

phagerna2 <- filter(rna2, gene_name == "T7p07")

gp1_plot <- ggplot(phagerna2, aes(x = strain, y = tpm)) + 
  stat_summary(geom="bar", fun.y = "mean", aes(group = strain, fill = strain)) +
  geom_point() +
  geom_line(aes(group = b_rep)) + 
  scale_fill_manual(values = strain_colors) +
  facet_grid(~ time) +
  ylab("relative mRNA abundance") +
  panel_border() +
  theme(legend.position = "none")

gp1_plot

save_plot("figures/gp1_rna.pdf", gp1_plot, base_aspect_ratio = 2)

gp1_10_plot <- plot_grid(gp10_plot, gp1_plot, ncol = 1, labels = c("A", "B"), align = "v")

save_plot("figures/gp1_10A_rna.pdf", gp1_10_plot, base_width = 8, base_height = 6)

```

The overall distributions between the two biological replicates look decent:

```{r fig.width=10, fig.height=8}

# Plot distribution with psuedo counts
ggplot(rna2 %>% mutate(tpm = tpm + 1), aes(x = tpm, fill = factor(b_rep))) + 
  geom_density(alpha = 0.5) + 
  scale_x_log10() + 
  facet_grid(strain ~ time)

```

### Using tRNA levels for normalization

tRNA content in each sample at each time:

```{r}
rna <- read_csv("./data/all_rna_no_ribo_trna_merge.csv") %>%
  # Replace strain names with something more useful
  mutate(strain = str_replace_all(strain, c("11-42" = "evol", "11-44" = "atten", "11-46" = "wt")))

phagerna <- rna %>%
  group_by(b_rep, strain, time, gene_type) %>%
  summarize(counts = sum(est_counts))

ribo_rna <- ggplot(phagerna %>% filter(org == "phage"), aes(x = strain, y = counts)) + 
  geom_bar(stat = "identity") + 
  facet_grid(b_rep ~ time)

ribo_rna
```

Normalizing to total tRNA content does not seem like an effective normalization strategy:

```{r}
# Normalize to total tRNAs

trnas <- rna %>%
  filter(gene_type == "tRNA") %>%
  group_by(b_rep, strain, time) %>%
  summarize(trna_count = sum(est_counts))

rna2 <- rna %>%
  left_join(trnas) %>%
  mutate(count = (est_counts/trna_count)*1e6)

ggplot(rna2 %>% filter(gene_name == "T7p45"), aes(x = strain, y = count, fill = factor(b_rep))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid(~ time)

```

