---
title: "Transcriptional Effects of T7 Attenuation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(readr)
library(dplyr)
library(stringr)
library(cowplot)

rna <- read_csv("./data/all_rna.csv") %>%
  # Replace strain names with something more useful
  mutate(strain = str_replace_all(strain, c("11-42" = "evol", "11-44" = "atten", "11-46" = "wt")))

```

## What proportion of transcripts are from T7?

At first glance, the abundance of T7 transcripts across strains and time points looks wildly different:

```{r fig.width = 9}

phagerna <- rna %>% 
  filter(org == "phage") %>%
  group_by(b_rep, time, strain) %>%
  summarize(tpm = sum(tpm))
  
ggplot(phagerna, aes(x = strain, y = tpm, fill = factor(b_rep))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid( ~ time)
```

These differences in TPM (transcripts per million), are driven by differences in rRNA levels:

```{r fig.width = 9}

phagerna <- rna %>%
  group_by(b_rep, strain, time, gene_type) %>%
  summarize(counts = sum(tpm))

ggplot(phagerna, aes(x = strain, y = counts, fill = factor(gene_type))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid(b_rep ~ time)

```

To account for different rRNA levels, I took the raw read counts and re-normalized them to the total mRNA content. This makes the phage transcript levels at least look more sane.

```{r fig.width = 9}
phagerna <- rna %>% 
  filter(gene_type == "CDS") %>%
  group_by(b_rep, strain, time) %>%
  mutate(total_count = sum(est_counts), est_counts_renorm = est_counts/total_count) %>% 
  group_by(b_rep, strain, time, org) %>%
  summarize(counts = sum(est_counts_renorm))
  
ggplot(phagerna, aes(x = strain, y = counts, fill = factor(b_rep))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid(org ~ time)
```

Lastly, lets look at gene 10A:

```{r}

phagerna <- filter(rna, gene_type == "CDS") %>%
  group_by(b_rep, strain, time) %>%
  mutate(total_count = sum(est_counts), est_counts_renorm = est_counts/total_count) %>%
  filter(gene_name == "T7p45")

ggplot(phagerna, aes(x = strain, y = est_counts_renorm, fill = factor(b_rep))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid(org ~ time)

```

...And T7 polymerase:

```{r}

phagerna <- filter(rna, gene_type == "CDS") %>%
  group_by(b_rep, strain, time) %>%
  mutate(total_count = sum(est_counts), est_counts_renorm = est_counts/total_count) %>%
  filter(gene_name == "T7p07")

ggplot(phagerna, aes(x = strain, y = est_counts_renorm, fill = factor(b_rep))) + 
  geom_bar(stat = "identity", position = "dodge") + 
  facet_grid(org ~ time)

```

Any individual genes still look somewhat noisy. This could have to do with the timing, or maybe a more sophisticated normalization method is required. Next, I plan to re-run kallisto with a reference transcriptome that doesn't contain any rRNA genes to see if kallisto's normalization is better than simply summing the estimated counts as I did above.
