---
title: "Phage Analyses"
output: html_document
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=FALSE, warning=FALSE, message=FALSE)

library(dplyr)
library(ggplot2)
library(cowplot)
library(readr)
library(stringr)
```


## Part 1: Proteomics Data

We have proteomics data from three different strains of T7 after 1, 5, and 9 minutes of infection in E. coli. Each time point has 2 replicates for a total of 18 samples.

**11-42**: evolved phage with codon-modified gene 10

**11-44**: initial codon-modified gene 10 phage

**11-46**: wild-type phage

**Normalization:** Data are areas normalized by total protein area (T7 and E. coli) for each sample.


### Replicate 1 shows biologically implausible abundances for major capsid protein in wildtype strain

```{r}

all_data <- read_csv('./rep1234_combined.csv', col_types = list(strain = col_character())) %>%
  filter(!is.na(class)) %>%
  mutate(strain = str_replace_all(strain, c("11-44" = "atten", "11-42" = "evol", "11-46" = "wt")))

# Correct wonky labeling
all_data <- mutate(all_data, time = ifelse(b_rep == 1 & time == 60 & strain == "wt", 1000, time)) %>% 
  mutate(time = ifelse(b_rep == 1 & time == 540 & strain == "wt", 60, time)) %>% 
  mutate(time = ifelse(b_rep == 1 & time == 1000 & strain == "wt", 540, time))

ggplot(filter(all_data, tabasco_id == 'gp10A', strain == "wt"), aes(x = factor(time), y = area_norm, group=b_rep, color=factor(b_rep))) + 
  geom_point(size = 3) +
  geom_line() + 
  panel_border() +
  ylab("relative protein abundance")


```
Samples from replicate 1 must have been mis-labeled, but it is impossible to determine the correct labeling. I suggest that we discard replicate 1. The remainder of the analyses here are presented from replicates 2 and 3. A 4th replicate is currently being processed.

### Major capsid protein levels are suppressed in attenuated and evolved strains

```{r}

all_data <- filter(all_data, org == "phage")

ggplot(filter(all_data, tabasco_id == 'gp10A'), aes(x = strain, y = area_norm, group = strain, color=factor(strain), order = desc(strain))) + 
  geom_point(size = 3) +
  stat_summary(geom="errorbar", fun.y="mean", aes(ymin=..y.., ymax=..y..), size=1) +
  facet_wrap(~ time) +
  panel_border() +
  ylab("relative protein abundance") +
  theme(legend.position = "none") +
  geom_text(aes(x = strain, y = area_norm, label = b_rep), hjust = -0.5)

```

### Phage proteins immediately downstream of gp10A are also suppressed

```{r, fig.height=20}

ggplot(filter(all_data, tabasco_id %in% c('gp10A','gp11','gp12','gp13','gp14','gp15')), aes(x = strain, y = area_norm, group = strain, color=factor(strain), order = desc(strain))) + 
  geom_point(size = 3) +
  stat_summary(geom="errorbar", fun.y="mean", aes(ymin=..y.., ymax=..y..), size=1) +
  facet_grid(tabasco_id ~ time, scales = "free_y") +
  panel_border() +
  ylab("relative protein abundance") +
  theme(legend.position = "none") +
  geom_text(aes(x = strain, y = area_norm, label = b_rep), hjust = -0.5)

```

### In aggregate, phage protein levels (not including major capsid) look about the same across strains

```{r}

make_abundance_plot <- function(protein_data) {

  sums <- group_by(protein_data, strain, time, b_rep) %>% 
    summarize(area_sum = sum(area_norm)) %>% 
    filter(!is.na(area_sum)) %>%
    group_by(strain, time) %>%
    summarize(area_mean = mean(area_sum), area_sd = sd(area_sum)) %>%
    mutate(upper_error = area_mean + area_sd, lower_error = area_mean - area_sd)

  myplot <- ggplot(sums, aes(x = strain, y = area_mean, group = strain, fill=factor(strain))) + 
    geom_bar(stat="identity") +
    geom_errorbar(aes(ymin = lower_error, ymax = upper_error), width = 0.5) +
    facet_wrap(~ time) +
    panel_border() +
    ylab('total relative protein abundance') +
    theme(legend.position = "none")
  
  return(myplot)

}

# plot_all_data <- make_abundance_plot(all_data)

plot_no10 <- make_abundance_plot(filter(all_data, 
                                        tabasco_id != 'gp10A', 
                                        tabasco_id != 'gp10B'))

plot_no10




# save_plot('../plot_all_data.pdf', plot_all_data, base_width = 10)
# save_plot('../plot_no10.pdf', plot_no10, base_width = 10)
# save_plot('../plot_class3_no10.pdf', plot_class3_no10, base_width = 10)
# save_plot('../plot_12.pdf', plot_class12, base_width = 10)
# 
# save_plot('../plot_all_datab.pdf', plot_all_datab, base_width = 8)
# save_plot('../plot_no10b.pdf', plot_no10b, base_width = 8)
# save_plot('../plot_class3_no10b.pdf', plot_class3_no10b, base_width = 8)
# save_plot('../plot_12b.pdf', plot_class12b, base_width = 8)

```


### Class 3 protein levels (not including capsid) look roughly the same

```{r}

plot_class3_no10 <- make_abundance_plot(filter(all_data, 
                                               tabasco_id != 'gp10A', 
                                               tabasco_id != 'gp10B', 
                                               class == 3))

plot_class3_no10

```

Averaged across all genes and all replicates for a given time point and given strain, class 3 protein levels look similar to gp10A (just barely). I have not actually tested to see if any strain is significantly different from another across all genes.


### Classes 1 and 2 also look roughly the same across all three strains

```{r}

plot_class12 <- make_abundance_plot(filter(all_data, class != 3))

plot_class12

```

### Some class 3 genes behave like gp10A, some do not

```{r, fig.height = 40}

ggplot(filter(all_data, tabasco_id != 'gp10A', class == 3), aes(x = strain, y = area_norm, group = strain, color=factor(strain))) + 
  geom_point(size = 3) +
  stat_summary(geom="errorbar", fun.y="mean", aes(ymin=..y.., ymax=..y..), size=1) +
  facet_grid(tabasco_id ~ time, scales='free_y') +
  panel_border() + 
  geom_text(aes(x = strain, y = area_norm, label = b_rep), hjust = -0.5)

```

Above is a detailed breakdown of every class 3 gene. Several genes follow the same pattern as gp10A, but others do not.