---
title: "Correlating proteomics and RNA-seq"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(broom)
library(purrr)
library(stringr)
library(cowplot)
library(gtools)

SAVE_PLOTS <- FALSE

# Set strain colors for all plots
strain_colors <- c(
  "recoded" = "orange",
  "evolved" = "lightblue",
  "wild-type" = "lightgreen"
)

load_rna <- function(path) {
  labels <- read_csv("../data/proteomics/id_map.csv") %>% mutate(accession = trimws(accession))

  rna <- read_csv(path) %>%
    # Replace strain names with something more useful
    mutate(strain = str_replace_all(strain, c("11-42" = "evol", "11-44" = "atten", "11-46" = "wt"))) %>%
    filter(gene_type != "tRNA") %>%
    mutate(tpm = tpm/1e6) %>%
    group_by(b_rep, time, strain) %>%
    mutate(tpm = tpm/sum(tpm)) %>% # re-normalize because of the small quantity of tRNAs
    filter(org == "phage") %>%
    group_by(strain, time, b_rep) %>%
    do(full_join(., labels, by = c("prot_id" = "accession")) %>% fill(strain, time, b_rep)) %>%
    ungroup() %>%
    mutate(gene_number = str_replace_all(gene_number, c("10A" = "10")), 
           time = as.numeric(str_replace_all(time, c("min" = "")))) %>%
    group_by(time, strain, gene_name, gene_number, prot_id, org) %>%
    summarize(avg_tpm = mean(tpm))
  
  rna
}

load_proteomics <- function(path) {
  labels <- read_csv("../data/proteomics/id_map.csv") %>% mutate(accession = trimws(accession))
  
  proteomics <- read_csv("../data/proteomics/abundances/all_reps.csv")

  proteomics <- proteomics %>%
    mutate(strain = str_replace_all(strain, c("11-44" = "atten", "11-42" = "evol", "11-46" = "wt"))) %>%
    mutate(protein_desc = trimws(protein_desc))
  
  # Join in standard T7 gene labels.
  proteomics <- filter(proteomics, org == "phage") %>% 
    group_by(strain, time, b_rep) %>%
    # Join by each group seperately so that missing phage proteins are filled in with NAs
    do(full_join(., labels, by = c("protein_desc" = "accession")) %>% fill(strain, time, b_rep)) %>%
    mutate(area_norm = ifelse(is.na(area_norm), 0, area_norm),
           gene_number = str_replace_all(gene_number, c("10A" = "10"))) # Fill in missing proteins (NAs) with 0
  
  mean_exp <- proteomics %>%
    group_by(time, strain, gene_number, class, protein_desc) %>% 
    summarize(area_norm = mean(area_norm))
  
  mean_exp
}
```

## Distribution of transcript abundances

```{r histograms, fig.width=16, fig.height = 10}

rna <- load_rna("../data/rna_seq/post_processed/all_rna_no_ribo_trna_merge.csv")

phage_rna <- rna %>% filter(strain != "evol", !is.na(avg_tpm), !is.na(gene_number))

rna_plot <- ggplot(phage_rna, aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), y = avg_tpm,
                                  fill = strain)) + 
  geom_bar(stat="identity", position="dodge") + 
  xlab("genes") + 
  ylab("relative transcript abundance") +
  theme(axis.text.x=element_text(angle=45,hjust=1))

rna_plot

if(SAVE_PLOTS) { save_plot("../figures/rna_plot.pdf", rna_plot, base_width=9, base_height = 5.5) }
```

## Correlation between proteomics and RNA seq
You can also embed plots, for example:

```{r pressure, echo=FALSE}
proteomics <- load_proteomics("../data/proteomics/abundances/all_reps.csv") %>% filter(strain != "evol")

protein_plot <- ggplot(proteomics %>% filter(!is.na(gene_number)), 
                       aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), 
                           y = area_norm, fill = strain)) + 
  geom_bar(stat="identity", position = "dodge") + 
  xlab("genes") + 
  ylab("relative protein abundance") +
  theme(axis.text.x=element_text(angle=45,hjust=1))

protein_plot

if(SAVE_PLOTS) { save_plot("../figures/protein_plot.pdf", protein_plot, base_width=9, base_height = 5.5) }

proteomics_rna <- full_join(proteomics, phage_rna, by = c("protein_desc" = "prot_id", 
                                                           "time" = "time", 
                                                           "strain" = "strain"))

proteomics_rna_9 <- filter(proteomics_rna, time == 9, avg_tpm != 0, area_norm != 0) %>% 
  group_by(strain) %>%
  mutate(avg_tpm = log(avg_tpm), area_norm = log(area_norm)) 

proteomics_rna_9_cor <- proteomics_rna_9 %>%
  nest() %>%
  mutate(cor = purrr::map(data, ~tidy(cor.test(.$area_norm, .$avg_tpm)))) %>%
  unnest(cor) %>%
  select(-data)

proteomics_rna_9_cor

protein_rna_plot <- ggplot(proteomics_rna_9, aes(x = area_norm, y = avg_tpm, color=strain)) + 
  geom_point() + 
  xlab("log(protein abundance)") + 
  ylab("log(transcript abundance)")

protein_rna_plot

if(SAVE_PLOTS) { save_plot("../figures/protein_rna_plot.pdf", protein_rna_plot) } 

```

# Correlating RNA-seq and simulated transcripts

```{r, fig.width = 16, fig.height= 10}

load_sim_data <- function(path) {
  raw_data <- read_tsv(path, col_names=F) 
  colnames(raw_data) <- c('time', 'gene', 'ribosomes', 'transcripts', 'density')
  labeled_data <- left_join(raw_data, labels, by = c("gene" = "pysinthe")) %>% 
    mutate(time = round(time)) # round times to whole seconds
}

sim_data <- load_sim_data("../data/simulated/run_072717_wt_density.tsv") 
  
bar_sim <- filter(sim_data, time %in% c(1200)) %>%
  left_join(labels) %>% filter(!is.na(gene_number))

bar_sim_plot <- ggplot(bar_sim, aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=transcripts)) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none") +
  labs(x="gene", y="transcript count") + theme(axis.text.x=element_text(angle=45,hjust=1))

bar_sim_plot

save_plot("../figures/sim_transcripts_9_plot.pdf", bar_sim_plot, base_width = 11, base_height = 3) 


sim_data_2 <- sim_data %>%
  filter(time == 1200) %>%
  left_join(phage_rna %>% filter(time == 9), by = c("gene_number" = "gene_number"))

ggplot(phage_rna, aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), y = avg_tpm)) + geom_bar(stat="identity") + facet_wrap(~time, ncol = 1)

cor.test(sim_data_2$avg_tpm, sim_data_2$transcripts)

```

```{r}

load_sim_data <- function(path) {
  raw_data <- read_tsv(path, col_names=F) 
  colnames(raw_data) <- c('time', 'gene', 'count')
  labeled_data <- left_join(raw_data, labels, by = c("gene" = "pysinthe")) %>% 
    mutate(time = round(time)) # round times to whole seconds
}

sim_data2 <- load_sim_data("../data/simulated/run_072717_wt_counts.tsv") 
  
bar_sim <- filter(sim_data2, time %in% c(1200)) %>%
  left_join(labels) %>% filter(!is.na(gene_number))

bar_sim_plot <- ggplot(bar_sim, aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=count)) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none") +
  labs(x="gene", y="protein count") + theme(axis.text.x=element_text(angle=45,hjust=1))

bar_sim_plot

save_plot("../figures/sim_proteins_9_plot.pdf", bar_sim_plot, base_width = 11, base_height = 3) 

```
```{r}

sim_data3 <- inner_join(sim_data, sim_data2, by=c("gene", "time")) %>%
  filter(count > 0) %>%
  group_by(time) %>%
  filter(n() > 6) %>%
  mutate(transcripts = log(transcripts), count = log(count)) %>%
  summarize(cor = cor.test(transcripts, count)$estimate, pval = cor.test(transcripts, count)$p.value)

cor_plot <- ggplot(sim_data3, aes(x = time, y = cor, color = -log(pval))) + geom_line() + ylim(-1,1) + geom_vline(xintercept = 1200)

save_plot("../figures/cor_plot.pdf", cor_plot, base_width = 6)

sim_data3 %>% filter(time == 1200) %>% print()

```