---
title: "Transcriptional Effects of T7 Attenuation"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE)

library(readr)
library(dplyr)
library(stringr)
library(cowplot)

```

## What proportion of transcripts are from T7?

### Mapping to a transcriptome without ribosomal RNA

I ran kallisto with a reference transcriptome that doesn't contain any rRNA genes. I also ran it with the experimentally derived mean fragment length of 450 and sd of 250. This causes short transcripts to drop out, but these are transcripts that can't be reliably quantitated given the fragment length.

Let's look at the overall proportion of transcripts between T7 and E. coli.

```{r}

strain_colors <- c(
  "recoded" = "orange",
  "evolved" = "lightblue",
  "wild-type" = "lightgreen"
)

rna2 <- read_csv("../data/rna_seq/post_processed/all_rna_no_ribo_trna_merge.csv") %>%
  # Replace strain names with something more useful
  mutate(strain = str_replace_all(strain, c("11-42" = "evolved", "11-44" = "recoded", "11-46" = "wild-type"))) %>%
  filter(gene_type != "tRNA") %>%
  mutate(tpm = tpm/1e6) %>%
  group_by(b_rep, time, strain) %>%
  mutate(tpm = tpm/sum(tpm)) # re-normalize because of the small quantity of tRNAs

rna2$strain <- factor(rna2$strain, levels = c("recoded", "evolved", "wild-type"))
rna2$time <- factor(rna2$time, labels = c("1" = "1 minute", "5" = "5 minutes", "9" = "9 minutes"))

phagerna2 <- rna2 %>% 
  group_by(b_rep, time, strain, org) %>%
  summarize(tpm = sum(tpm))

phagerna2$org <- factor(phagerna2$org, labels = c("ecoli" = "E. coli", "phage" = "T7"))
  
rna_no_ribo <- ggplot(phagerna2, aes(x = strain, y = tpm, group=strain)) + 
  stat_summary(geom="bar", fun.y="mean", aes(fill = strain)) +
  scale_fill_manual(values = strain_colors) +
  geom_point(aes(group=b_rep)) +
  geom_line(aes(group=b_rep)) +
  facet_grid(org ~ time) +
  panel_border() +
  ylab("relative mRNA abundance") +
  theme(legend.position = "none")

rna_no_ribo

save_plot("../figures/rna_no_ribo.pdf", rna_no_ribo, base_height = 5, base_width = 8)
```

Gene 10A transcripts and gene 1 transcripts:

```{r}
phagerna2 <- filter(rna2, gene_name == "T7p45")

gp10_plot <- ggplot(phagerna2, aes(x = strain, y = tpm, group = b_rep)) + 
  stat_summary(geom="bar", fun.y = "mean", aes(group = strain, fill = strain)) +
  geom_point() +
  geom_line(aes(group = b_rep)) + 
  scale_fill_manual(values = strain_colors) +
  facet_grid(~ time) +
  ylab(expression(atop("relative mRNA abundance", paste("(gene ", italic("10A"), ")")))) +
  panel_border() +
  theme(legend.position = "none")

gp10_plot

phagerna2 <- filter(rna2, gene_name == "T7p07")

gp1_plot <- ggplot(phagerna2, aes(x = strain, y = tpm)) + 
  stat_summary(geom="bar", fun.y = "mean", aes(group = strain, fill = strain)) +
  geom_point() +
  geom_line(aes(group = b_rep)) + 
  scale_fill_manual(values = strain_colors) +
  facet_grid(~ time) +
  ylab(expression(atop("relative mRNA abundance", paste("(gene ", italic("1"), ")")))) +
  panel_border() +
  theme(legend.position = "none")

gp1_plot

gp1_10_plot <- plot_grid(gp10_plot, gp1_plot, ncol = 1, labels = c("A", "B"), align = "v")

save_plot("../figures/gp1_10A_rna.pdf", gp1_10_plot, base_width = 8, base_height = 6)

```
