---
title: "Simulating the T7 Infection Cycle"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE, message=FALSE)
library(tidyverse)
library(stringr)
library(cowplot)
library(forcats)
library(gtools)
library(ggrepel)

# Add some defaults to theme
theme_cowplot <- theme_cowplot() + 
  background_grid(major = 'y', minor = "none") + # add thin horizontal lines 
  panel_border()

theme_set(theme_cowplot)

# Helper simulation data loading function
labels <- read_csv("../data/proteomics/id_map.csv")
load_sim_data <- function(path) {
  raw_data <- read_tsv(path, col_names=F) 
  colnames(raw_data) <- c('time', 'gene', 'count')
  labeled_data <- left_join(raw_data, labels, by = c("gene" = "pysinthe")) %>% 
    mutate(time = round(time)) # round times to whole seconds
}
```

## Simulated data

The following simulation data come from a wild-type simulation of a T7 infection about ~25C, run for a total of 20 minutes of infection time. This plot shows all proteins over the course of the infection.

```{r load_sim_data}

wildtype_sim <- load_sim_data("../data/simulated/run_072717_wt_counts.tsv")

```

```{r line_plot_sim, fig.width=10}

line_sim <- filter(wildtype_sim, !is.na(class))

line_sim_plot <- ggplot(line_sim, aes(x=time, y=count, group=gene, color=factor(class))) +
  geom_line() +
  labs(x="time (s)", y="protein count", color="class") +
  panel_border(remove=TRUE)

line_sim_plot
# save_plot("line_sim.pdf", line_sim_plot, base_width = 9, base_height = 5)

```

```{r bar_sim_plot, fig.width=15, fig.height=15}
bar_sim <- filter(wildtype_sim, !is.na(class), time %in% c(400, 800, 1200, 1600, 2000)) %>%
  left_join(labels)

bar_sim_plot <- ggplot(bar_sim, aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=count, fill=factor(class))) +
  geom_bar(stat = "identity") +
  facet_wrap(~time, scales = "free_x", ncol = 1) +
  theme(legend.position = "none") +
  labs(x="gene", y="count")

bar_sim_plot

```

```{r density_plot}

# density_data <- wildtype_sim %>%
#   filter(str_detect(gene, "rnapol*")) %>%
#   select(time, gene, count) %>%
#   spread(gene, count, fill = 0) %>%
#   mutate(transcribing = `rnapol-1_total` - `rnapol-3.5` - `rnapol-1`)
# 
# rnapol_density_plot <- ggplot(density_data, aes(x = time, y = transcribing)) +
#   geom_line() + 
#   labs(x = "time (s)", y = "T7 RNApol density (counts/genome)") + xlim(0, 1260)

# save_plot("~/Desktop/density_plot.pdf", rnapol_density_plot, base_aspect_ratio = 1.4)

```

### Ribosome densities, wildtype

```{r ribo_density_plot, fig.width = 10, fig.height = 10}

ribo_data <- read_tsv("../data/simulated/run_072717_wt_density.tsv", col_names = c("time", "gene", "ribosomes", "transcripts", "density"))

ggplot(ribo_data, aes(x = time, y = density, group = gene, color = gene)) + 
  geom_line() +
  geom_text_repel(
    data = ribo_data %>% filter(time == max(time)),
    aes(label = gene),
    size = 4,
    nudge_x = 45,
    segment.color = "grey80"
  ) +
  theme(legend.position = "none") +
  coord_cartesian(xlim = c(min(ribo_data$time), max(ribo_data$time) + 100))

```

### Ribosome densities on gene 10, wildtype and recoded

```{r ribo_density_plot2, fig.width = 10, fig.height = 10}

ribo_data_10 <- read_tsv("../data/simulated/run_072717b_rec_density.tsv", col_names = c("time", "gene", "ribosomes", "transcripts", "density")) %>%
  filter(gene == "gene 10A") %>% mutate(condition = "recoded")

ribo_data_10_wild <- read_tsv("../data/simulated/run_072717_wt_density.tsv", col_names = c("time", "gene", "ribosomes", "transcripts", "density")) %>%
  filter(gene == "gene 10A") %>% mutate(condition = "wildtype")

all_ribo <- bind_rows(ribo_data_10, ribo_data_10_wild)

ggplot(all_ribo, aes(x = time, y = density, group = condition, color = condition)) + 
  geom_line()

```

## Comparing simulations to experimental data

We have proteomics data for a wild-type T7 infected _E. coli_ at 1, 5, and 9 minutes post-infection. These experiments were conducted at 37C (?), while most rate constants used in the simulation were derived from experiments conducted at 25C or 30C. At higher temperatures the infection cycle is shorter, so 9 minutes at 37C is comparable to 20 minutes at 25C. Thus we will compare the 9 minute time-point experimental data to the 20 minute time-point simulated data.

```{r load_exp_data}
all_exp <- read_csv("../data/proteomics/abundances/all_reps.csv")

all_exp <- all_exp %>%
  mutate(strain = str_replace_all(strain, c("11-44" = "atten", "11-42" = "evol", "11-46" = "wt"))) %>%
  mutate(protein_desc = trimws(protein_desc))

# Join in standard T7 gene labels.
all_exp <- filter(all_exp, org == "phage") %>% 
  group_by(strain, time, b_rep) %>%
  # Join by each group seperately so that missing phage proteins are filled in with NAs
  do(full_join(., labels, by = c("protein_desc" = "accession")) %>% fill(strain, time, b_rep)) %>%
  mutate(area_norm = ifelse(is.na(area_norm), 0, area_norm)) # Fill in missing proteins (NAs) with 0

mean_exp <- all_exp %>% filter(strain == "wt") %>%
  group_by(time, gene_number, class) %>% 
  summarize(area_norm = mean(area_norm))
```

### All genes experimental (9 minutes) and simulated (20 minutes)

The following plot shows experimental protein counts 9 minutes after infection (top) and simulated protein counts (bottom)

```{r late_sim_exp_plot, fig.height=6.5, fig.width=15}
late_sim <- filter(wildtype_sim, time == 1200) %>% 
  rename(sim_time = time) # rename to avoid conflicts when joining exp data
  
late_mean_exp <- filter(mean_exp, time == 9)

late_sim_exp <- full_join(late_mean_exp, late_sim) %>% 
  rename(exp_count = area_norm, sim_count = count) %>%
  select(sim_time, gene_number, class, exp_count, sim_count) %>%
  gather(source, count, exp_count:sim_count) %>%
  mutate(count = ifelse(is.na(count), 0, count)) %>%
  filter(!is.na(class))

sim_exp_plot <- ggplot(late_sim_exp, aes(x=factor(gene_number, levels=unique(mixedsort(gene_number))), y=count, fill=factor(class))) +
  geom_bar(stat = "identity") +
  facet_wrap(~factor(source, levels=c("sim_count", "exp_count"), labels=c("simulated", "experimental")), scales = "free", ncol = 1) +
  theme(legend.position = "none") +
  labs(x="gene", y="protein count")

# save_plot("sim_plot.pdf", sim_exp_plot, base_width = 8, base_height = 5)

sim_exp_plot

```

<!-- ### In our experiments, gene 10 was detected after 1 minute of infection -->

```{r all_exp_plot, fig.height=10, fig.width=15}

# all_exp_plot <- ggplot(all_exp %>% filter(!is.na(class), strain=="wt"), aes(x=factor(gene_number, levels=unique(mixedsort(gene_number))), y=area_norm, fill=factor(class))) +
#   geom_bar(stat="identity") +
#   facet_grid(time~b_rep, scales="free_x") +
#   labs(x="gene", y="relative abundance") +
#   theme(legend.position="none")
# 
# all_exp_plot
```

<!-- #### Here is the same plot as above, but now the scales have been adjusted to show differences between genes -->

<!-- Why do the distribution of genes look so similar across all three time points? -->

```{r all_exp_scaled_plot, fig.height=10, fig.width=15}

# all_exp_plot + facet_grid(time~b_rep, scales="free")

```

<!-- ### And now excluding replicate 4... -->

```{r all_exp_scaled_plot_corrected, fig.height=10, fig.width=15}

# all_exp_no4 <- all_exp_plot %+% 
#   filter(all_exp, !is.na(class), b_rep != 4, strain == "wt") + 
#   facet_grid(time~b_rep, scales="free_x")
# 
# all_exp_no4
# 
# all_exp_no4 + facet_grid(time~b_rep, scales="free")

```

## Genomic manipulations _in silico_

<!-- - `wildtype`: wildtype T7 simulation -->
<!-- - `slow_pol`: the speed of T7 RNA polymerase was reduced from 240bp/s to ~100bp/s -->
<!-- - `promoter_knockout`: all early T7 promoters were knocked out except for phi1.3, which is required for the full genome to enter the cell -->

```{r genome_manip_plots, fig.height=9, fig.width=15}

recoded_sim <- load_sim_data("../data/simulated/run_072717b_rec_counts.tsv") %>% mutate(condition="recoded")
wildtype_sim <- load_sim_data("../data/simulated/run_072717_wt_counts.tsv") %>% mutate(condition="wildtype")
all_sim <- bind_rows(wildtype_sim, recoded_sim) %>% filter(time == 1200)

ggplot(all_sim %>% filter(!is.na(class)), aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=count, fill=factor(class))) +
  geom_bar(stat="identity") +
  facet_wrap(~fct_relevel(condition, "wildtype", "recoded"), ncol = 1, scales="free_x") +
  labs(x="gene", y="count") +
  theme(legend.position="none")

```