---
title: "Correlating proteomics and RNA-seq"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
library(tidyverse)
library(forcats)
library(broom)
library(purrr)
library(stringr)
library(cowplot)
library(gtools)
library(ggrepel)

cbPalette <- c("#555555", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

SAVE_PLOTS <- TRUE

# Set strain colors for all plots
strain_colors <- c(
  "recoded" = "orange",
  "evolved" = "lightblue",
  "wild-type" = "lightgreen"
)

load_rna <- function(path) {
  labels <- read_csv("../../data/id_map.csv") %>% mutate(accession = trimws(accession))

  rna <- read_csv(path) %>%
    # Replace strain names with something more useful
    mutate(strain = str_replace_all(strain, c("11-42" = "evol", "11-44" = "atten", "11-46" = "wt"))) %>%
    filter(gene_type != "tRNA") %>%
    mutate(tpm = tpm/1e6) %>%
    group_by(b_rep, time, strain) %>%
    mutate(tpm = tpm/sum(tpm)) %>% # re-normalize because of the small quantity of tRNAs
    filter(org == "phage") %>%
    group_by(strain, time, b_rep) %>%
    do(full_join(., labels, by = c("prot_id" = "accession")) %>% fill(strain, time, b_rep)) %>%
    ungroup() %>%
    mutate(gene_number = str_replace_all(gene_number, c("10A" = "10")), 
           time = as.numeric(str_replace_all(time, c("min" = "")))) %>%
    group_by(time, strain, gene_name, gene_number, prot_id, org) %>%
    summarize(avg_tpm = mean(tpm))
  
  rna
}

load_proteomics <- function(path) {
  labels <- read_csv("../../data/id_map.csv") %>% mutate(accession = trimws(accession))
  
  proteomics <- read_csv("../../data/proteomics.csv")

  proteomics <- proteomics %>%
    mutate(strain = str_replace_all(strain, c("11-44" = "atten", "11-42" = "evol", "11-46" = "wt"))) %>%
    mutate(protein_desc = trimws(protein_desc))
  
  # Join in standard T7 gene labels.
  proteomics <- filter(proteomics, org == "phage") %>% 
    group_by(strain, time, b_rep) %>%
    # Join by each group seperately so that missing phage proteins are filled in with NAs
    do(full_join(., labels, by = c("protein_desc" = "accession")) %>% fill(strain, time, b_rep)) %>%
    mutate(area_norm = ifelse(is.na(area_norm), 0, area_norm),
           gene_number = str_replace_all(gene_number, c("10A" = "10"))) # Fill in missing proteins (NAs) with 0
  
  mean_exp <- proteomics %>%
    group_by(time, strain, gene_number, class, protein_desc) %>% 
    summarize(area_norm = mean(area_norm))
  
  mean_exp
}

```

## Distribution of transcript abundances

```{r histograms, fig.width=16, fig.height = 6}

rna <- load_rna("../../data/transcripts.csv")

phage_rna <- rna %>% filter(strain != "evol", !is.na(gene_number), !is.na(prot_id)) %>% mutate(avg_tpm = if_else(!is.na(avg_tpm), avg_tpm*10000, 0))

rna_plot <- ggplot(phage_rna, aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), y = avg_tpm,
                                  fill = fct_relevel(strain, "wt", "atten"))) + 
  geom_bar(stat="identity", position="dodge") + 
  xlab("gene") + 
  ylab("transcript abundance") +
  # theme_minimal() +
  theme(# axis.text.x=element_text(angle=45,hjust=1), 
        legend.position = c(0.9, 0.85),
        axis.text.x=element_blank()) +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual("strain", values = cbPalette, labels = c("wild type", "attenuated")) 

proteomics <- load_proteomics("../../data/proteomics.csv") %>% filter(strain != "evol", !is.na(protein_desc)) %>% mutate(area_norm = area_norm*100000)

protein_plot <- ggplot(proteomics %>% filter(!is.na(gene_number)), 
                       aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), 
                           y = area_norm, fill = fct_relevel(strain, "wt", "atten"))) + 
  geom_bar(stat="identity", position = "dodge") + 
  xlab("gene") + 
  ylab("protein abundance") +
  theme(# axis.text.x=element_text(angle=45,hjust=1),
        axis.text.x=element_blank(), 
        legend.position = c(0.9, 0.85)) +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual("strain", values = cbPalette, labels = c("wild type", "attenuated"))

protein_rna_bar_plot <- plot_grid(rna_plot, protein_plot, ncol = 1)

if(SAVE_PLOTS) { save_plot("../../figures/protein_rna_bar_plot.pdf", protein_rna_bar_plot, base_width=9, base_height = 5) }

proteomics_rna <- inner_join(proteomics, phage_rna, by = c("protein_desc" = "prot_id", 
                                                           "time" = "time", 
                                                           "strain" = "strain"))

proteomics_rna_9 <- filter(proteomics_rna, time == 9, !is.na(class), avg_tpm > 0, area_norm > 0) %>% 
  group_by(strain) %>%
  mutate(avg_tpm = log10(avg_tpm), area_norm = log10(area_norm)) %>% filter(avg_tpm > 1, area_norm > 1) 

proteomics_rna_9_cor <- proteomics_rna_9 %>%
  mutate(class = if_else(class == 1, as.numeric(2), as.numeric(class))) %>%
  group_by(strain, class) %>%
  nest() %>%
  mutate(cor = purrr::map(data, ~tidy(cor.test(.$area_norm, .$avg_tpm)))) %>%
  unnest(cor) %>%
  select(-data)

proteomics_rna_9_cor

proteomics_rna_9_cor2 <- proteomics_rna_9 %>%
  group_by(strain) %>%
  nest() %>%
  mutate(cor = purrr::map(data, ~tidy(cor.test(.$area_norm, .$avg_tpm)))) %>%
  unnest(cor) %>%
  select(-data)

proteomics_rna_9_cor2

```

```{r}

protein_rna_scatter_both <- ggplot(proteomics_rna_9, aes(y = area_norm, x = avg_tpm, color=fct_relevel(strain, "wt", "atten"))) + 
  geom_line(aes(group=gene_name), color = "grey80") +
  geom_point(size = 2) + 
  xlab("log10(transcript abundance)") + 
  ylab("log10(protein abundance)") +
  xlim(1, 3.6) + ylim(1, 3.6) +
  theme(legend.position = c(0.15, 0.85)) +
  scale_color_manual("strain", values = cbPalette, labels = c("wild type", "attenuated"))

protein_rna_scatter_both

protein_rna_scatter_wt_class <- ggplot(proteomics_rna_9 %>% filter(strain == "wt"), aes(y = area_norm, x = avg_tpm, color=factor(class))) + 
  geom_point(size = 2) + 
  xlab("log10(transcript abundance)") + 
  ylab("log10(protein abundance)") +
  xlim(1, 3.6) + ylim(1, 3.6) +
  scale_color_manual("class", values = cbPalette[c(3, 4, 7)]) +
  theme(legend.position = c(0.1, 0.85))

protein_rna_scatter_wt_class

protein_rna_scatter_wt <- ggplot(proteomics_rna_9 %>% filter(strain == "wt"), aes(y = area_norm, x = avg_tpm)) + 
  geom_point(color = cbPalette[1], size = 2) + 
  xlab("log10(transcript abundance)") + 
  ylab("log10(protein abundance)") +
  xlim(1, 3.6) + ylim(1, 3.6)

protein_rna_scatter_wt

protein_rna_scatter_labels <- protein_rna_scatter_both + 
  geom_text_repel(data=proteomics_rna_9 %>% mutate(gene_number.x = if_else(strain == "atten", "", gene_number.x)), 
                  aes(avg_tpm, area_norm, label = gene_number.x),
                  box.padding = unit(0.7, 'lines'),
                  point.padding = unit(0.7, 'lines'),
                  color = 'black',
                  segment.alpha = 0.2,
                  force = 0.5) 

protein_rna_scatter_labels

if(SAVE_PLOTS) { 
  save_plot("../../figures/protein_rna_scatter_both.pdf", protein_rna_scatter_both, base_aspect_ratio = 1.4)
  save_plot("../../figures/protein_rna_scatter_wt.pdf", protein_rna_scatter_wt, base_aspect_ratio = 1.4)
  save_plot("../../figures/protein_rna_scatter_wt_class.pdf", protein_rna_scatter_wt_class, base_aspect_ratio = 1.4)
  save_plot("../../figures/protein_rna_scatter_labels.pdf", protein_rna_scatter_labels, base_aspect_ratio = 1.4)
} 

```

# Correlating RNA-seq and simulated transcripts

```{r, fig.width = 16, fig.height= 40}

load_simulation <- function(path) {
  labels <- read_csv("../../data/id_map.csv") %>% mutate(accession = trimws(accession))
  raw_counts <- read_tsv(paste0(path, "_counts.tsv"), col_names=F) 
  colnames(raw_counts) <- c('time', 'gene', 'proteins')
  raw_densities <- read_tsv(paste0(path, "_density.tsv"), col_names = F)
  colnames(raw_densities) <- c('time', 'gene', 'ribosomes', 'transcripts', 'density')
  raw_data <- full_join(raw_counts, raw_densities, by = c("time", "gene"))
  labeled_data <- left_join(raw_data, labels, by = c("gene" = "pysinthe")) %>%
    mutate(time = round(time)) %>%
    mutate(proteins = if_else(is.na(proteins), 0.0, as.numeric(proteins)),
           transcripts = if_else(is.na(transcripts), 0.0, as.numeric(transcripts)),
           ribosomes = if_else(is.na(ribosomes), 0.0, as.numeric(ribosomes))) %>%
    filter(!is.na(gene_number)) %>%
    select(-accession)
}

sim_data_wt <- data_frame(path = paste0("T7_082017_w", seq(1,20)), weight = seq(1, 20)) %>%
  mutate(data = map(path, ~load_simulation(paste0("../../runs/", .)))) %>%
  mutate(condition = "wt") %>%
  unnest()

sim_data_atten <- data_frame(path = paste0("T7_082017_w", seq(1,20), "_r"), weight = seq(1, 20)) %>%
  mutate(data = map(path, ~load_simulation(paste0("../../runs/", .)))) %>%
  mutate(condition = "atten") %>%
  unnest()

sim_data <- bind_rows(sim_data_wt, sim_data_atten)

# test <- load_simulation("../runs/run_081417c_wt")

# sim_data <- load_multiple_simulations(c("../runs/run_081417b_rec",
#                                         "../runs/run_081417b_wt",
#                                         "../runs/run_081417_wt",
#                                         "../runs/run_081417_rec"),
#                                       c("rec", "wt", "wt", "rec"),
#                                       c("7x", "7x", "20x", "20x"))

# sim_data <- load_multiple_simulations(c("../runs/run_081417b_rec",
#                                         "../runs/run_081417b_wt",
#                                         "../runs/run_081417_wt",
#                                         "../runs/run_081417_rec",
#                                         "../runs/run_081617_wt"),
#                                       c("rec", "wt", "wt", "rec", "wt"),
#                                       c(7, 7, 20, 20, 1))
  
bar_sim <- filter(sim_data, time == 1200)

bar_sim_plot <- ggplot(bar_sim, aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=proteins, fill=condition)) +
  geom_bar(stat = "identity", position="dodge") +
  facet_wrap(~weight, ncol = 1, scales = "free") +
  labs(x="gene", y="protein count") + theme(axis.text.x=element_text(angle=45,hjust=1)) +
  scale_fill_manual(values = cbPalette)

bar_sim_plot

bar_sim_plot_density <- ggplot(bar_sim, aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=density, fill=condition)) +
  geom_bar(stat = "identity", position="dodge") +
  facet_wrap(~weight, ncol = 1, scales = "free_x") +
  labs(x="gene", y="ribosome density") + theme(axis.text.x=element_text(angle=45,hjust=1)) +
  scale_fill_manual(values = cbPalette)

bar_sim_plot_density

save_plot("../../figures/sim_transcripts_9_plot.pdf", bar_sim_plot, base_width = 11, base_height = 30) 

```

```{r, fig.width=16, fig.height = 30}

bar_sim_plot_transcript <- ggplot(bar_sim, aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=transcripts, fill=condition)) +
  geom_bar(stat = "identity", position="dodge") +
  facet_wrap(~weight, ncol = 1, scales = "free") +
  labs(x="gene", y="transcript count") + theme(axis.text.x=element_text(angle=45,hjust=1)) +
  scale_fill_manual(values = cbPalette)

bar_sim_plot_transcript

cor.test(bar_sim$transcripts, bar_sim$proteins)

# save_plot("../figures/sim_proteins_9_plot.pdf", bar_sim_plot, base_width = 11, base_height = 3) 

```

```{r}

protein_rna_sim_scatter <- ggplot(sim_data %>% filter(time == 1200, gene_number != 1), 
       aes(y = log10(proteins), x = log10(transcripts), color=fct_relevel(condition, "wt", "atten"), group=gene_number)) +
  geom_line(color = "grey80") +
  geom_point() +
  scale_color_manual("strain", values = cbPalette) +
  xlim(1, 3.5) + ylim(2, 3.8) +
  facet_wrap(~weight, scales = "free_y")

protein_rna_sim_scatter_wt <- ggplot(sim_data %>% filter(time == 1200, gene_number != 1, weight == 11, condition=="wt"), 
       aes(y = log10(proteins), x = log10(transcripts))) +
  geom_point() +
  xlim(1, 3.5) + ylim(2, 3.8) +
  scale_color_manual(values = cbPalette)

protein_rna_sim_scatter_wt_class <- ggplot(sim_data %>% filter(time == 1200, gene_number != 1, weight == 11, condition=="wt"), 
       aes(y = log10(proteins), x = log10(transcripts), color=factor(class))) +
  geom_point() +
  xlim(1, 3.5) + ylim(2, 3.8) +
  scale_color_manual("class", values = cbPalette)

protein_rna_sim_scatter_labeled <- protein_rna_sim_scatter + 
  geom_text_repel(data=sim_data %>% filter(time == 1200, gene_number != 1, weight == 11) %>% mutate(gene_number = if_else(condition == "atten", "", gene_number)), 
                  aes(log10(transcripts), log10(proteins), label = gene_number),
                  box.padding = unit(0.7, 'lines'),
                  point.padding = unit(0.7, 'lines'),
                  color = 'black',
                  segment.alpha = 0.2,
                  force = 0.5)

protein_rna_sim_scatter

protein_rna_sim_scatter_wt

protein_rna_sim_scatter_wt_class

protein_rna_sim_scatter_labeled

if(SAVE_PLOTS) { save_plot("../../figures/protein_rna_sim_scatter.pdf", protein_rna_sim_scatter, base_width = 15, base_height = 10) } 
  

```


### 20x

```{r}
library(viridis)

gene_list <- unique(proteomics_rna_9$gene_number.x)

sim_data_cors <- filter(sim_data, proteins > 0, transcripts > 0, condition=="wt", gene_number != 1, gene_number != 0.7, gene_number != 3.5, gene_number != 2, gene_number %in% gene_list) %>%
  mutate(class = if_else(class == 1, 2, as.numeric(class))) %>%
  group_by(condition, weight, time, class) %>%
  filter(n() > 3) %>%
  mutate(transcripts = log10(transcripts), proteins = log10(proteins)) %>%
  summarize(cor = cor.test(transcripts, proteins)$estimate, pval = cor.test(transcripts, proteins)$p.value)

cor_plot <- ggplot(sim_data_cors %>% filter(weight == 3), aes(x = time, y = cor, group=factor(class), color=factor(class))) + 
  geom_hline(yintercept = 0) +
  geom_line() + 
  ylim(-1,1) + xlim(400, 1400) +
  theme(legend.position = c(0.92, 0.25), legend.background = element_rect(fill="white"), panel.grid.major.y = element_line(colour="grey80")) +
  scale_color_manual("class", values = cbPalette[c(4,7)], labels = c("1 and 2", "3"))

cor_plot

sim_data_cors2 <- filter(sim_data, proteins > 0, transcripts > 0, condition=="wt", gene_number != 1, gene_number != 0.7, gene_number != 3.5, gene_number != 2, gene_number %in% gene_list) %>%
  group_by(condition, weight, time) %>%
  filter(n() > 3) %>%
  mutate(transcripts = log10(transcripts), proteins = log10(proteins)) %>%
  summarize(cor = cor.test(transcripts, proteins)$estimate, pval = cor.test(transcripts, proteins)$p.value)

cor_plot2 <- ggplot(sim_data_cors2 %>% filter(weight == 3), aes(x = time, y = cor)) + 
  geom_hline(yintercept = 0) +
  geom_line() + 
  xlim(400, 1400) + 
  ylim(-1,1) + theme(panel.grid.major.y = element_line(colour="grey80"))

cor_plot2

cor_plots_all <- plot_grid(cor_plot2, cor_plot, ncol = 1, align = "v")

save_plot("../../figures/cor_plot.pdf", cor_plots_all, base_width = 6, base_height = 6)

# save_plot("../../figures/cor_plot_by_class.pdf", cor_plot, base_aspect_ratio = 2.5)
# save_plot("../../figures/cor_plot.pdf", cor_plot2, base_aspect_ratio = 1.8)

sim_data_cors %>% filter(time == 1200) %>% print()

```
