---
title: "RNA-seq Re-analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
library(tidyverse)
library(ggplot2)
library(gtools)
library(stringr)
library(cowplot)
```

## Tools

Reads were mapped with HISAT2 and counted with bedtools multicov. Counts were normalized and given as TPM values.

## Normalized counts show full coverage of genome

In the plot below, reads are counted towards all genes that they map to if a read overlaps with multiple genes.

```{r, echo=FALSE, fig.width = 10, fig.height = 18, fig.width = 12}

import_counts <- function(file_name) {
  labels <- read_csv("output/id_map.csv") %>% mutate(accession = trimws(accession))
  read_tsv(file_name, col_names=FALSE) %>%
  filter(X3 == "CDS") %>%
  select("start" = X4, "end" = X5, "gene" = X9, "count" = X10) %>%
  mutate(gene_number = str_match(gene, "Note=(possible )?gene ([0-9\\.\\-AB]*)")[,3],
         lengthkb = (end - start)/1000,
         rpk = count/lengthkb,
         scale_factor = sum(rpk)/1000000,
         tpm = rpk/scale_factor) %>%
    left_join(labels, by = c("gene_number" = "gene_number"))
} 

rep1_5 <- import_counts("data/rna_seq/PA196_T7W-R1-05m_S27_L004_R1_001_sort_counts.tsv") %>% mutate(b_rep = 1, time = 5)
rep1_10 <- import_counts("data/rna_seq/PA198_T7W-R1-10m_S28_L004_R1_001_sort_counts.tsv") %>% mutate(b_rep = 1, time = 10)
rep1_15 <- import_counts("data/rna_seq/PA200_T7W-R1-15m_S29_L004_R1_001_sort_counts.tsv") %>% mutate(b_rep = 1, time = 15)
rep1_20 <- import_counts("data/rna_seq/PA202_T7W-R1-20m_S30_L004_R1_001_sort_counts.tsv") %>% mutate(b_rep = 1, time = 20)
rep1_25 <- import_counts("data/rna_seq/PA204_T7W-R1-25m_S31_L004_R1_001_sort_counts.tsv") %>% mutate(b_rep = 1, time = 25)


all_reps <- bind_rows(rep1_5, rep1_10, rep1_15, rep1_20, rep1_25) %>% mutate(strain = "wildtype")

ggplot(all_reps, aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), y = tpm)) + 
  geom_bar(stat="summary", fun.y = "mean", aes(fill = factor(class))) +
  theme(axis.text.x=element_text(angle=45,hjust=1), legend.position = "none") +
  xlab("gene") +
  facet_wrap(~time, ncol = 1, scales = "free")
plot_5 <- ggplot(rep1 %>% filter(time == 5, !is.na(class)), aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), y = tpm)) + 
  geom_bar(stat="summary", fun.y = "mean", aes(fill = factor(class))) +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.9)) +
  scale_fill_discrete("class") +
  xlab("gene") + ylab("transcripts per million") + scale_y_continuous(expand = c(0,0))
# save_plot("../../figures/transcripts_5min.pdf", plot_5, base_width = 12, base_height = 5)

# write_csv(all_reps, "../../data/transcripts_remapped.csv")
```

### Simulating a terminator T3.8

```{r}

load_simulation <- function(path) {
  labels <- read_csv("../../data/id_map.csv") %>% mutate(accession = trimws(accession))
  raw_counts <- read_tsv(paste0(path, "_counts.tsv"), col_names=F) 
  colnames(raw_counts) <- c('time', 'gene', 'proteins')
  raw_densities <- read_tsv(paste0(path, "_density.tsv"), col_names = F)
  colnames(raw_densities) <- c('time', 'gene', 'ribosomes', 'transcripts', 'density')
  raw_data <- full_join(raw_counts, raw_densities, by = c("time", "gene"))
  labeled_data <- left_join(raw_data, labels, by = c("gene" = "pysinthe")) %>%
    mutate(time = round(time)) %>%
    mutate(proteins = if_else(is.na(proteins), 0.0, as.numeric(proteins)),
           transcripts = if_else(is.na(transcripts), 0.0, as.numeric(transcripts)),
           ribosomes = if_else(is.na(ribosomes), 0.0, as.numeric(ribosomes))) %>%
    filter(!is.na(gene_number))
}

sim_data_wt <- load_simulation("../../runs/012418/run_T7_012418_wt")

bar_sim <- filter(sim_data_wt, time %in% c(300, 400, 450, 500, 1100), !is.na(gene_number))

bar_sim_plot <- ggplot(bar_sim, aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=transcripts)) + geom_bar(stat="identity") + facet_wrap(~time, ncol = 1, scales = "free_y")

bar_sim_plot

sim_plot_9 <- ggplot(bar_sim %>% filter(time == 1100), aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=transcripts, fill=factor(class))) + 
  geom_bar(stat="identity") + 
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.9)) +
  scale_fill_discrete("class") +
  xlab("gene") + ylab("simulated transcripts") + scale_y_continuous(expand = c(0,0)) 

kallisto <- read_csv("../../data/transcripts.csv")

kal_plot <- ggplot2(kallisto %>% filter(strain == "wt"))

sim_exp_9 <- plot_grid(plot_5, sim_plot_5, ncol = 1, align="v")
# save_plot("../../figures/old_sim_exp_9min.pdf", sim_plot_9, base_width=10, base_height = 4)
# save_plot("../../figures/old_sim_exp_9min_transcripts.pdf", sim_exp_9, base_width = 10, base_height = 8)
```

### Simulation and real transcripts side-by-side

```{r}

sim_final <- filter(sim_data_wt, time == 1100)
sim_genes <- unique(sim_final$gene_number)

sim_plot <-ggplot(sim_final, aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=transcripts, fill= factor(class))) + 
  geom_bar(stat="identity") + xlab("gene") + ylab("simulated transcripts") +
  theme(axis.text.x=element_text(angle=45,hjust=1), legend.position = c(0.0, 0.85)) +
  scale_fill_discrete(name="class")

exp_final <- filter(all_reps, time == 9, gene_number %in% sim_genes)

exp_plot <- ggplot(exp_final, aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), y = tpm, fill = factor(class))) + 
  geom_bar(stat="summary", fun.y = "mean") +
  geom_point(alpha=0.5) +
  theme(axis.text.x=element_text(angle=45,hjust=1), legend.position = c(0.0, 0.85)) +
  scale_color_discrete("replicate") +
  xlab("gene") + ylab("transcripts per million") +
  scale_fill_discrete(name="class")

sim_exp_plot <- plot_grid(exp_plot, sim_plot, cols = 1)

save_plot("T7_rna_exp_sim.pdf", sim_exp_plot, base_width = 10, base_height = 8)

sim_plot <-ggplot(sim_final, aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=transcripts, fill= factor(class))) + 
  geom_bar(stat="identity") + xlab("gene") + ylab("simulated transcripts") +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.85)) +
  scale_fill_discrete(name="class") + scale_y_continuous(expand = c(0, 0))

exp_plot <- ggplot(exp_final, aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), y = tpm, fill = factor(class))) + 
  geom_bar(stat="summary", fun.y = "mean") +
  geom_point(alpha=0.5) +
  theme(axis.text.x=element_blank(), axis.title.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.85)) +
  scale_color_discrete("replicate") +
  ylab("transcripts per million") +
  scale_fill_discrete(name="class") + scale_y_continuous(expand = c(0, 0))

sim_exp_plot <- plot_grid(exp_plot, sim_plot, cols = 1, align = "v")

save_plot("T7_rna_exp_sim_no_labels.pdf", sim_exp_plot, base_width = 8, base_height = 6)
```

### Proteomics and rna correlations

```{r}

load_proteomics <- function(path) {
  labels <- read_csv("../../data/id_map.csv") %>% mutate(accession = trimws(accession))
  
  proteomics <- read_csv("../../data/proteomics.csv")

  proteomics <- proteomics %>%
    mutate(strain = str_replace_all(strain, c("11-44" = "atten", "11-42" = "evol", "11-46" = "wt"))) %>%
    mutate(protein_desc = trimws(protein_desc))
  
  # Join in standard T7 gene labels.
  proteomics <- filter(proteomics, org == "phage") %>% 
    group_by(strain, time, b_rep) %>%
    # Join by each group seperately so that missing phage proteins are filled in with NAs
    do(full_join(., labels, by = c("protein_desc" = "accession")) %>% fill(strain, time, b_rep)) %>%
    mutate(area_norm = ifelse(is.na(area_norm), 0, area_norm),
           gene_number = str_replace_all(gene_number, c("10A" = "10"))) # Fill in missing proteins (NAs) with 0
  
  mean_exp <- proteomics %>%
    group_by(time, strain, gene_number, class, protein_desc) # %>% 
    # summarize(area_norm = mean(area_norm))
  
  mean_exp
}

proteomics <- load_proteomics("") %>% filter(time == 9, strain == "wt", area_norm != 0)
phage_rna <- all_reps %>% filter(time == 9) %>% group_by(gene_number) %>% summarize(avg_tpm = mean(tpm)/1000000) %>% filter(avg_tpm != 0)
proteomics_rna <- inner_join(proteomics %>% group_by(gene_number, class) %>% summarize(area_norm = mean(area_norm)), phage_rna, by = c("gene_number" = "gene_number"))

exp_cor <- ggplot(proteomics_rna, aes(x = log10(avg_tpm), y = log10(area_norm), color = factor(class))) + 
  geom_point() + 
  xlab("log10(transcript abundance)") + 
  ylab("log10(protein abundance)") + 
  scale_color_discrete(name = "class")

simulated_cor <- ggplot(sim_final %>% filter(transcripts > 20, proteins > 20), aes(x = log10(transcripts), y = log10(proteins), color = factor(class))) + 
  geom_point() +
  xlab("log10(simulated transcript abundance)") + 
  ylab("log10(simulated protein abundance)") + 
  scale_color_discrete(name = "class")


scatter_plots <- plot_grid(exp_cor, simulated_cor, rows = 1)

save_plot("scatter_plots.pdf", scatter_plots, base_aspect_ratio = 2.5)

exp_prot_final <- proteomics %>% filter(!is.na(gene_number)) %>% ungroup() %>% mutate(gene_number = str_replace(gene_number, "10", "10A"))
prot_genes <- unique(exp_prot_final$gene_number)

sim_plot_prot <-ggplot(sim_final %>% filter(gene_number %in% prot_genes), aes(x=factor(gene_number, unique(mixedsort(gene_number))), y=proteins, fill= factor(class))) + 
  geom_bar(stat="identity") + xlab("gene") + ylab("simulated protein abundance") +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.85)) +
  scale_fill_discrete(name="class") + scale_y_continuous(expand = c(0, 0))

exp_plot_prot <- ggplot(exp_prot_final, aes(x = factor(gene_number, levels=unique(mixedsort(gene_number))), y = area_norm, fill = factor(class))) + 
  geom_bar(stat="summary", fun.y = "mean") +
  geom_point(alpha=0.5) +
  theme(axis.text.x=element_blank(), axis.title.x=element_blank(), axis.ticks.x=element_blank(), legend.position = c(0.0, 0.85)) +
  scale_color_discrete("replicate") +
  ylab("relative protein abundance") +
  scale_fill_discrete(name="class") + scale_y_continuous(expand = c(0, 0))

sim_exp_plot_prot <- plot_grid(exp_plot_prot, sim_plot_prot, cols = 1, align = "v")

save_plot("T7_prot_exp_sim_no_labels.pdf", sim_exp_plot_prot, base_width = 8, base_height = 6)


```

### Looking at individual mapped reads

```{r, fig.width=12, }

coverage <- read_tsv("../../data/PA54_coverage_q60.tsv", col_names = c("genome", "position", "coverage"))
full_map <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 24162) + xlim(0,40000)
save_plot("../../figures/cov_map.pdf", full_map, base_width = 10, base_height = 5)
plot1 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 24210) + geom_vline(xintercept = 24162) + xlim(24090,24300)
plot2 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 11161) + geom_vline(xintercept = 11203) + xlim(11000,11300)
zoomed_cov_map <- plot_grid(plot1, plot2)
save_plot("../../figures/zoomed_cov_map.pdf", zoomed_cov_map, base_width = 9, base_height = 4)

rnase_1 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 3138) + xlim(2000,4000)
rnase_1_1 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 5887) + xlim(5500,6500)
rnase_1_3 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 6448) + xlim(6000,7000)
rnase_4_7 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 13892) + xlim(13700,14000)
rnase_6_5 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 18563) + xlim(18300,18700)
rnase_13 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 27281) + xlim(27000,28000)
rnase_18_5 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 36856) + xlim(36000,37000)
plot2 <- ggplot(coverage, aes(x = position, y = coverage)) + geom_bar(stat="identity") + geom_vline(xintercept = 11161) + geom_vline(xintercept = 11203) + xlim(11000,11300)
zoomed_cov_map <- plot_grid(plot1, plot2)

zoomed_cov_map_rnase <- plot_grid(rnase_4_7, rnase_6_5)
save_plot("../../figures/zoomed_cov_map_rnase.pdf", zoomed_cov_map_rnase, base_width = 9, base_height = 4)

```